src/Compilers/Core/Portable/Operations/BasicBlock.cs

```diff
@@ -1,152 +1,119 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Linq;
 using Microsoft.CodeAnalysis.PooledObjects;
 
-namespace Microsoft.CodeAnalysis.Operations
+namespace Microsoft.CodeAnalysis.FlowAnalysis
 {
-    /// <summary>
-    /// PROTOTYPE(dataflow): Add documentation
-    /// </summary>
-    public enum BasicBlockKind
+    internal partial class ControlFlowGraphBuilder
     {
-        Entry,
-        Exit,
-        Block
-    }
-
-    /// <summary>
-    /// PROTOTYPE(dataflow): Add documentation
-    /// PROTOTYPE(dataflow): We need to figure out how to split it into a builder and 
-    ///                      a public immutable type.
-    /// </summary>
-    public sealed class BasicBlock
-    {
-        private readonly ImmutableArray<IOperation>.Builder _statements;
-        private readonly ImmutableHashSet<BasicBlock>.Builder _predecessors;
-        internal (IOperation Value, Branch Branch) InternalNext;
-        internal (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
-
-        public BasicBlock(BasicBlockKind kind)
-        {
-            Kind = kind;
-            _statements = ImmutableArray.CreateBuilder<IOperation>();
-            _predecessors = ImmutableHashSet.CreateBuilder<BasicBlock>();
-        }
-
-        public BasicBlockKind Kind { get; private set; }
-        public ImmutableArray<IOperation> Statements => _statements.ToImmutable();
-
-        /// <summary>
-        /// PROTOTYPE(dataflow): Tuple is temporary return type, we probably should use special structure instead.
-        /// </summary>
-        public (IOperation Condition, bool JumpIfTrue, Branch Branch) Conditional => InternalConditional;
-
         /// <summary>
-        /// PROTOTYPE(dataflow): During CR there was a suggestion to use different name - "Successor".
+        /// PROTOTYPE(dataflow): Rename this type to "BasicBlockBuilder" and source file to "ControlFlowGraphBuilder.BasicBlockBuilder.cs"
         /// </summary>
-        public (IOperation Value, Branch Branch) Next => InternalNext;
-
-        public ImmutableHashSet<BasicBlock> Predecessors => _predecessors.ToImmutable();
+        internal sealed class BasicBlock
```

Comment:
Can we name this something else, like `BasicBlockBuilder` or `BasicBlockInternal`? I don't want to have two different types with the same name that interact constantly, that will be a big source of confusion imo.

Reply:
Yes, covered by the prototype comment above, which we will address in a follow up PR. We avoided doing so in this PR to avoid too many merge conflicts in the CFG builder.

---
In reply to: [192246644](https://github.com/dotnet/roslyn/pull/27254#discussion_r192246644) [](ancestors = 192246644)